platform-sdk/swirlds-platform-core/src/main/java/com/swirlds/platform/Browser.java

```diff
@@ -636,12 +636,47 @@ private Collection<SwirldsPlatform> createLocalPlatforms(
                 final SignedState loadedSignedState = getUnmodifiedSignedStateFromDisk(
                         mainClassName, swirldName, nodeId, appVersion, addressBook.copy(), emergencyRecoveryManager);
 
+                // check software version compatibility
+                final SoftwareVersion loadedSoftwareVersion = loadedSignedState == null
```

Comment:
I think this series of checks should be extracted to a static utility method, either here or in a utility class. That will make this method shorter and easier to read, and the static method can easily be unit tested.

```
static boolean detectSoftwareUpgrade(final SoftwareVersion appVersion, final SignedState loadedSignedState)
```

Reply:
The detectSoftwareUpgrade() method has been added to BootstrapUtils.java